pkg/apis/sessions/session_state.go

```diff
@@ -59,78 +65,119 @@ func (s *SessionState) String() string {
-			&ss.RefreshToken,
-		} {
-			err := into(s, c.Encrypt)
-			if err != nil {
-				return "", err
-			}
+// EncodeSessionState returns an encrypted, lz4 compressed, MessagePack encoded session
+func (s *SessionState) EncodeSessionState(c encryption.Cipher, compress bool) ([]byte, error) {
+	// Marshal to MessagePack
+	packed, err := msgpack.Marshal(s)
+	if err != nil {
+		return nil, err
+	}
+
+	if !compress {
+		// Encrypt the msgpack encoded
+		return c.Encrypt(packed)
+	}
+
+	// Compress the packed encoding
+	// The Compress:Decompress ratio is 1:Many. LZ4 gives fastest decompress speeds
+	buf := new(bytes.Buffer)
+	zw := lz4.NewWriter(nil)
+	zw.Header = lz4.Header{
+		BlockMaxSize:     65536,
+		CompressionLevel: 0,
+	}
+	zw.Reset(buf)
+
+	reader := bytes.NewReader(packed)
+	_, err = io.Copy(zw, reader)
+	if err != nil {
+		return nil, err
+	}
+	err = zw.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	compressed, err := ioutil.ReadAll(buf)
+	if err != nil {
+		return nil, err
+	}
+
+	// Encrypt the compressed
+	return c.Encrypt(compressed)
+}
+
+// DecodeSessionState decodes a LZ4 compressed MessagePack into a Session State
+func DecodeSessionState(data []byte, c encryption.Cipher, compressed bool) (*SessionState, error) {
+	// Decrypt
+	decrypted, err := c.Decrypt(data)
+	if err != nil {
+		return nil, err
+	}
+
+	packed := decrypted
+	if compressed {
+		// LZ4 Decompress
+		reader := bytes.NewReader(decrypted)
+		buf := new(bytes.Buffer)
+		zr := lz4.NewReader(nil)
+		zr.Reset(reader)
+		_, err = io.Copy(buf, zr)
+		if err != nil {
+			return nil, err
+		}
+
+		packed, err = ioutil.ReadAll(buf)
+		if err != nil {
+			return nil, err
 		}
 	}
 
-	b, err := json.Marshal(ss)
-	return string(b), err
+	// Decode MessagePack
+	var ss *SessionState
+	err = msgpack.Unmarshal(packed, &ss)
+	if err != nil {
+		return nil, err
+	}
+
+	return ss, nil
 }
 
-// DecodeSessionState decodes the session cookie string into a SessionState
-func DecodeSessionState(v string, c encryption.Cipher) (*SessionState, error) {
+// LegacyV5DecodeSessionState decodes a legacy JSON session cookie string into a SessionState
+func LegacyV5DecodeSessionState(v string, c encryption.Cipher) (*SessionState, error) {
 	var ss SessionState
 	err := json.Unmarshal([]byte(v), &ss)
 	if err != nil {
 		return nil, fmt.Errorf("error unmarshalling session: %w", err)
 	}
 
-	if c == nil {
-		// Load only Email and User when cipher is unavailable
-		ss = SessionState{
-			Email:             ss.Email,
-			User:              ss.User,
-			PreferredUsername: ss.PreferredUsername,
-		}
-	} else {
-		// Backward compatibility with using unencrypted Email or User
-		// Decryption errors will leave original string
-		err = into(&ss.Email, c.Decrypt)
-		if err == nil {
-			if !utf8.ValidString(ss.Email) {
-				return nil, errors.New("invalid value for decrypted email")
-			}
+	// Backward compatibility with using unencrypted Email or User
+	// Decryption errors will leave original string
```

Comment:
Do you think we should drop this now, since we are dropping support for unencrypted sessions in the next release? Everyone should be upgrading via v6 so this should be redundant I think

Reply:
Let's do it! -- I have no qualms about using this as the opportunity to purge unencrypted sessions.  We should roll this & https://github.com/oauth2-proxy/oauth2-proxy/pull/575 together then (after V6 gives a rollout window of encrypted sessions + SHA256 HMAC)

That would likely make the forced reauth for stragglers cleanest and result in no potential errors since their session will be tossed out with the SHA1 cookie failing to validate.